### ISSUE TITLE
Generate Location-specific pdf forms from the template form

### ISSUE THREAD

**Issue by bsaiken** [FWAJL/FieldWorkAssistantMVC#778]
This Issue generates Location-Specific pdf forms from a pdf template form.

Navigate to: `/task/listAll`
For this Issue, there is only one Task in the example.  However, there may be many Tasks listed, so the code must select for the selected Task.
When the button `class="btn btn-warning from-inactive-list"` is clicked, several server-side functions will be called.  Once function is to generate Location-specific forms.  In the example, use the Task "PDF Task" (it is the only option).
If you navigate to `/task/locations`, you will observe there are 6 Locations (PDF Location 1, PDF Location 2, etc.).
If you navigate to `/task/forms`, you will see there is only one form, "Well Purge".  This form is to be used as a template to create 6 new, Location-specific forms, which are copies of the "Well Purge" form.

**Comment by bsaiken** [comment]
@mehedee1 @FWAJL 

I realize we can't start this, and other Issues until after Monday.  I need to discuss with Jeremie.

While you could create the copies of the Location-specific forms, you will need to place the forms somewhere.  The table that is used to identify the form, the form type, and the url has not been created yet.   In addition, we need to identify the `task_id` that the form is related to.  I am not certain what directory the form would get saved in.  The only forms we have handled so far are forms we upload, so they are in the /upload/ directory.  However, the forms you will be generating are really being "uploaded" so we may want to store them somewhere else.  

Jeremie...?

**Comment by mehedee1** [comment]
/task/forms
>>>
1. When these forms are viewed, all updated data will be populated and then pop up pdf form? 
2. What is the relation between Well Purge and location 1, 2, 3...... 
3. Also what is the data map of Well Purge?

**Comment by bsaiken** [comment]
@mehedee1 
1. 
In this first step, all you are doing is creating the copies of the master form.  No viewing or populating yet.
So, there will be three new files when you are finished, and they will be stored in Web/uploads/location_form
The three files will need unique file names, so the file names will be:
Well Purge_PDF Location 1_PDF Task.pdf
Well Purge_PDF Location 2_PDF Task.pdf
Well Purge_PDF Location 3_PDF Task.pdf
or generically, `title`_`location_name`_`task_name`

2.
 The number of Locations determine how many copies you will make, and each location is part of the new form file name.  I do not know the field name for "Well ID" on the "Well Purge" pdf master file. This is the field where the location_name will go later.  But, Do NOT insert the location_name yet.

3.
I do not know what you mean by data map.

**Comment by mehedee1** [comment]
1. when pdf files will be copied? clicking on arrow button of task/listAll from inactive to active? What happen if it goes from active to inactive?

2. Right /task/forms 
Right now only one form Well Purge. If there would be many, the copied would be many? Example:
Well Purge_PDF Location 1_PDF Task.pdf
Well Purge_PDF Location 2_PDF Task.pdf
Well Purge_PDF Location 3_PDF Task.pdf

Chain Of Custody_PDF Location 1_PDF Task.pdf
Chain Of Custody_PDF Location 2_PDF Task.pdf
Chain Of Custody_PDF Location 3_PDF Task.pdf

My concept is clear?

3. I see database:
a. form_id in master_form
b. master_form_id in task_form 
so master form id determines which file in /task/forms. Am I right?

4. task_location will determine how many template will be copied as Location_x. Am I right?

**Comment by bsaiken** [comment]
1. Navigate to Task > List All.  You will see there is only one Task, "PDF Task".  This Task is inactive.  If you click the Task, then click the arrow button, you will get a popup confirmation box asking, "Sure to Activate this Task?"  It is after the user clicks "OK" that the pdf files are copied.
It is not your job to worry about what happens when the file goes from active to inactive.  We are just taking this one step at a time.

2. There is only one Well Purge form because this form is a template.  The number of copies depends on the number of Locations.  If you look at the /task/locations tab, you will see there are only three Task Locations, the three I identified.  So, there will be three copies of the master form.

Only work on the form "Well Purge".  It is the only task form.  If you navigate to /task/forms, you will see only "Well Purge".  The Chain of Custody form is Issue #785.  We are just taking this one step at a time.

1. Yes, master_form_id in the task_form table tells you which form to copy.

2. Yes, task_location determines how many copies are made.  One form for each Location.

**Comment by mehedee1** [comment]
Done!
When arrow button is clicked task form will be copied as per location.

**Comment by bsaiken** [comment]
@mehedee1 
This is a great start.  I realize we need to change some things to be consistent with our document handling parameters and to make the file names unique in all situations. 
I have a meeting tomorrow morning with Jeremie, so I will give you my revisions after that. Also, we have made some significant changes to the site that I will want to migrate over to Blue Host.

This is why it is very important for you to carefully document the code that you have added and make a backup on your local server.  Your work will get over-written when we update the Bluehost site - I don't want you to have to start from scratch!

More tomorrow.

**Comment by mehedee1** [comment]
I have kept backup.

**Comment by bsaiken** [comment]
@mehedee1 
I talked to Jeremie about our file naming structure and how to generate the copies.  

When you generate a new file, we have already prepared a method to generate the file name, populate the relation table, and identify the directory (or createt he directory, if it has not been created yet).  

You would call the method `UploadFiles` in `FileUploader.php`.
This would automatically fill in the `document` table.  The `document.document_category` is going to be `task_location_id`, and the file name would then be `task_location_id__guid.pdf`, where guid is always going to have a unique value.  The table entry for `document.document_title` would be "title"_"Location Name", where "title" is the title of the master form.

Since we have not built all the document handling methods, it makes sense to use them.  

Is this something you could do?  If not, I will have someone else make the copies and then have you modify the pdf files once they are created.

**Comment by mehedee1** [comment]
I will see the method. If get confused I will ask.

**Comment by mehedee1** [comment]
To copy files, I joined several table and create file name ("title"_"location_name"_"task_name"). You can see /includes/location_copy.php

Would you tell me where the FileUploader.php is?

**Comment by bsaiken** [comment]
@mehedee1 
The filename "location_name"_"task"name" will not be the file name, see above.

I just tested this and I did not see that any new files were created.

The code in location_copy.php does not populate the `document` table.

The correct code structure is in FileUploader.php:
Library/Core/Utility/FileUploader.php

**Comment by mehedee1** [comment]
Getting problem using existed fileupload method. It will be better to be done by someone. After copying with this method, let me know. I will then imply pdf action.

**Comment by bsaiken** [comment]
@gsouvik 
See Issue #727.
Under your login name, I have created a Task "Example Task for Issue #778", which I will use for my examples. 
The Task has three locations, `task_location_id`= 302, 303 and 304.
So, when you activate this task, there will be three new forms:
`302_guid1.pdf`
`303_guid2.pdf`
`304_guid3.pdf`
And these will be stored in the directory `/uploads/task-location`.  You will also need to update the `document` table, too.

Please let me know if you have any questions.

**Comment by FWAJL** [comment]
@gsouvik @bsaiken 

TO complete this task, you will need to use the FileUploader class in Library/Core/Utility. It is built for uploading files but I believe you can create a controller that builds all the arrays required to built a "copy" file method.
Then the step of saving the physical file will be different in the sense that you will copy existing files to another location with a different name.

Let us know if you need help.

**Comment by gsouvik** [comment]
@bsaiken @FWAJL 
I read the thread from top till the last message and mostly the requirement is clear to me, however I have few questions regarding the setup which I would like to ask:
1) I don't have a local directory under the project root (or Web) which reads `uploads`; neither I could find the template PDF files which actually needs to be copied into location specific PDF files. Am I missing anything.
2) I understand that location specific PDFs would be stored into the `document` table. Which `id` or data would actually replace the string `guid` in the file name? Or in other words what would replace `guid` (example provided by you: `302_guid1.pdf`) when the script copies the files? Or is it handled by the already built methods I would be using? 
3) Once the location specific PDFs are generated, where are those displayed in the frontend, or is it out of scope of this issue and I need not have to worry about that? (part of #727 ?)

I think if I have the physical template files on my local and have the above queries explained, I'm ready to start. 

Thank you.

**Comment by bsaiken** [comment]
@gsouvik 
1) The local directory is automatically generated.  I uploaded the pdf file to eLance.  If you login using the credentials admintest:test, you can upload the file, and then observe how everything works.  Navigate to `Filed Forms > Add master forms`
Then you will need to promote the form to the Task.  
Go to `Filed Forms > List All` and promote the "Well Purge" form from the left to right column.
Then
Go to `Task > List All`, right click the Task you want to edit (add a Task, if you do not have one one locally), then choose the `Forms` tab.  You might need to check the box that says "Task requires completion of field form" (you wrote that code) - :-)
Then promote the form again from left to right column.
Now, when you activate the Task, you will have your template form ready to be copied.
2) The guid is automatically generated by the already built methods..
3) Yes, that is Issue 727.

**Comment by FWAJL** [comment]
@bsaiken @gsouvik 

In additional to Brian's answer, please have a read of the document: https://docs.google.com/document/d/1zBHTfTBSp6MEK6ItFB1Ws3W9JYWpzGdDODQj8NjYauw

It explains quickly what goes on in the document upload that we will use similarly to copy forms. The only difference is that since you don't upload files but copy them, the global variable $_FILES is not filled out. 
Therefore, you will need to fill it out with the data you need to make copies of forms. The logic happens in the FileController (Library/Controllers) and the DocumentDal (Library/DAL/Modules/Models) for the upload.
The trick is to use the Upload example to make the copies. 

It is complex but I think you can manage it. Let us know if you need more help.

**Comment by gsouvik** [comment]
@FWAJL 
I've nearly done the coding but having problem with just one thing. Can you please provide me the steps by which I could test the method
`executeUpload`
of the `FileController` class on my local computer. I want to see what data is passed into the variable `dataPost`.

Thank you.

**Comment by FWAJL** [comment]
@gsouvik 

All is in [this view module](https://github.com/FWAJL/FieldWorkAssistantMVC/blob/dev/Applications/PMTool/Views/Modules/upload_file.php) at lines 6 and 7.

On other note, what IDE do you use? NetBeans? Eclipse?

**Comment by gsouvik** [comment]
@FWAJL 
I already noticed it's in that view module, but I DONT know where that view module get's included and in which route I can get it.

Can you please copy paste the link of the page from your local which uses that view module. 
I use Eclipse and Sublime text, may I know why you wish to know?

Thanks.

**Comment by FWAJL** [comment]
@gsouvik @damirius 

Go to http://dev.fieldworkmanager.com/technician/showForm?mode=edit&technician_id=1 to view the upload photo example.

Please read this and hopefully, it is clear enough for you: https://docs.google.com/document/d/1jLMfE0X370rFGxqRYw3sLca38mrYaLnvhjjaSN0rwfc
If not, please ask Damir for help as I won't be available much this weekend.

Finally, I am using NetBeans and it has a great debugger that allows to work efficiently. Maybe Eclipse does too.

Thank you.

**Comment by damirius** [comment]
@gsouvik

You'll need few things to post to file/upload (executeUpload action) along with the file. 
"itemCategory" should contain string with primary key of the linked table. For example on technician page we upload technician pictures to `document` table also, but we need to link them somehow to the `technician` table, so we use "technician_id" as the "itemCategory" there, it's called `document_category` in the `document` table.
Next one is "itemId" that's the id of the record we are linking this document to. So if we set "technician_id" as the itemCategory, we can set itemId to 1, to link it to the technician with primary key set to 1.
There are few others, "itemReplace" can be true or false, if it is set to true it will first try to find document record with same itemCategory and itemId to delete it first. We use that for technicians since they can have only 1 picture.
Other than that we have "title" which is used for image captions or readable names, since we hash the filenames and of course "file".
Like Jeremie mentioned we use "upload_file.php" module along with dropzone.js for easier file upload and ability for users to drag&drop files.

**Comment by gsouvik** [comment]
@damirius @FWAJL  
Thank you both for the input. It really helped. 
In between I've figured out how the Document model (DAO) has to be populated. Next I understand that we are not actually uploading a file but copying an existing file into different ones. But the requirement states we have to use the existing code of `execudeUpload`. As @FWAJL suggested I'm simulating the $_FILES array which that method receives when actually a file is uploaded. For my code with the master form, file array looks like:
`Array
(
    [file] => Array
        (
            [name] => 051d827b-123c-4edb-82a4-d800abce2ed2.pdf
            [type] => pdf
            [tmp_name] => /Web/uploads/master_form/051d827b-123c-4edb-82a4-d800abce2ed2.pdf
            [error] => 0
            [size] => 152
        )

)`

But I never see the file actually uploaded. I think `tmp_name` is expected to contain the temporary file name of the file on the server and not a file on the webroot. So this is actually not supposed to work. That brings me to @FWAJL 's suggestion of simulating $_FILES. I'm not sure how we can do that. 

I think if we can simply write the guid generation code into a small function, this task could be completed in minutes. Just an opinion though. Looking forward to further suggestions.

Thanks again.

**Comment by damirius** [comment]
@gsouvik

Ah, I see. So if I got it correctly, when you activate the task on the task list some template form should be copied into multiple pdfs depending on the number of locations in the activated task right?

If that's the case, we can reuse some of the code of `executeUpload` action in the `TaskController->executeUpdateItems` action, since if I'm correct that's the one which changes the task from active to inactive and vice versa. 
You can use this part:
`$manager = $this->managers()->getManagerOf("Document");
    $manager->setRootDirectory($this->app()->config()->get(\Library\Enums\AppSettingKeys::RootDocumentUpload));
    $manager->setWebDirectory($this->app()->config()->get(\Library\Enums\AppSettingKeys::BaseUrl) . $this->app()->config()->get(\Library\Enums\AppSettingKeys::RootUploadsFolderPath));`
That will set up proper root and web directories for the manager to use.
After that you can set up `$directory` to `task_location`.
Other part of the code is put inside `foreach ($task_ids as $id)`:
`$manager->setFilenamePrefix($id.'_');
    $document = new \Applications\PMTool\Models\Dao\Document();
    $document->setDocument_category('task_location_id');
    $document->setDocument_title('GENERATED TITLE');
   `
Next part is the problem `$manager->addWithFile($document,$files['file']);`
`addWithFile` is the method in the DocumentDal base manager class. You can find it in Library/Dal/Modules/DocumentDal.php.
You can create similar one which can be used for copying. For example addWithExistingFile($document,'source_document_id');
So you can fetch there existing document record and use that info to create new pdf records.

**Comment by gsouvik** [comment]
@damirius 
Exactly my thoughts. That should be the best approach. I'll definitely let you know if I run into any more problems.

And very true, move_uploaded_file would always look into the temp folder of the server, it won't work with regular webroot files.

Cheers!

**Comment by gsouvik** [comment]
@bsaiken 
Finally I've been able to reuse the maximum amount of available code and copy the master template into location specific PDF file.

Please note I haven't added any validation if the user deactivates the task and activates it back again. In that case currently multiple PDFs are generated again. Also as the guid is always auto generated and a new string every time, all the PDF names are different and cannot be tracked for if it is already generated in the previous activation process.

Please review and let me know and I shall make any changes if required. 

Thank you.

**Comment by bsaiken** [comment]
@gsouvik 
Deactivating Tasks will not be possible once the site is more fully developed.  So don't worry about Task deactivation and re-activation.

I see that the task forms are created.  There is one other thing to fix.  The `document_title` needs to be modified.  The table entry for `document.document_title` would be "title"_"Location Name", where "title" is the `document_title` of the master form.  This makes the `document_title`, which is displayed to the user, unique and easily identifiable.  You added `_x` after `pdf` - do not do that, and do not include the extension. 

All administrative master forms are going to have the title format `FWMx-titletext.pdf`.  When you use an administrative master form, only use the `titletext` as the document title.

So, as an example, if the Master form title is `FWM_L-WellPurge.pdf` and you have two Locations, `Location 1` and `Location 2`, the titles of the the copied files should be:
`WellPurge_Location 1`
`WellPurge_Location 2`

**Comment by gsouvik** [comment]
@bsaiken 
But on my local that's exactly what is happening. Please see the screenshot of the document table after I executed the script.

The task location names were `MW-5B` and `MW-6B` and I'm using the `document.title` field for the master form title, which in my case reads `Well Purge 2`.

The only thing which I might have missed is the processing of `FWMx-titletext.pdf` to `titletext` part on `document.title`. Is that what you referring to?

![screen shot 2015-04-18 at 9 43 27 pm](https://cloud.githubusercontent.com/assets/2292482/7216223/1cc2fe4a-e615-11e4-8ce5-43319f08ca05.png)

**Comment by bsaiken** [comment]
@gsouvik 
I see now.   You do not need to process `FWMx-titletext.pdf`.  My mistake.

Now you can move on to listing them!

**Comment by bsaiken** [comment]
@FWAJL 
Jeremie,
Once you this code is ready for release, will you [please update the Bluehost site (don't forget the database).  Mehedee is ready to populate the pdf fields, but we need this file copy feature working for him to continue.

**Comment by FWAJL** [comment]
@gsouvik @bsaiken 

At first look, the code is working fine. Good job completing it @gsouvik :+1: . And thank you to @damirius for helping.

However the code is not doing what we need and we querying the database when we should query the session data. 

1. In `CreateLocationSpecificPDF`of `TaskHelper`, at lines 289-291, we only copy the first template of `$template_data` for each task location whereas we should copy all the templates of `$template_data`. A for loop is therefore required to process all forms.
2. In the for loop from line 303, you look up the location object by querying the database. You find the location in Session at this stage using `FindIndexInObjectListById` in the `CommonHelper` that takes the id value, the property id name (location_id), the sessionArray (Project), and the sessionKey (ProjectLocations).

Let me know if there is any question.

**Comment by gsouvik** [comment]
@FWAJL 
Following changes are made:
1) If there are multiple master templates associated with a task, all of them are copied into location specific PDFs.
2) Instead of querying the database for location, it is now fetched from session using the `FindIndexInObjectListById` method.

Please review and let me know if this looks right.

**Comment by bsaiken** [comment]
@gsouvik @FWAJL 
I am confused about the status of this Issue.

See my comment on Issue #778

**Comment by gsouvik** [comment]
@bsaiken @FWAJL 
Now that #727 is fixed, what should be status of this issue.

**Comment by bsaiken** [comment]
@FWAJL @gsouvik 
I agree, so I am closing this Issue.  We can always re-open it or create a new Issue if need be.

**Comment by FWAJL** [comment]
@bsaiken 

As I said, I see a bug. I will look into it as I think it is something I did.

**Comment by bsaiken** [comment]
@WebDevJL 
I tested this on the DEV branch and it works.  I will leave it to you to close this Issue.

**Comment by FWAJL** [comment]
@bsaiken 

Actually, it was released in version 1.6.21 so you can test it. I did and it works but a second person testing is always good. Close the issue if you don't find another bug.

Thank you.
